File: 241011/싸움땅/battle-ground.py
class Main:
    def __init__(self):
        self.n, self.m, self.k = map(int, input().split())
        self.grid = [[-1] * (self.n + 2)] + \
                    [[-1] + list(map(int, input().split())) + [-1] for _ in range(self.n)] + \
                    [[-1] * (self.n + 2)]
        self.players = [[0] * 4] + [list(map(int, input().split())) for _ in range(self.m)]
        self.points = [0] * (self.m + 1)
        self.gun = [0] * (self.m + 1)
        self.gun_grid = [[[] for _ in range(self.n + 2)] for _ in range(self.n + 2)]
        self.directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    def init_grid(self):
        for y in range(1, self.n + 1):
            for x in range(1, self.n + 1):
                if self.grid[y][x] > 0:
                    self.gun_grid[y][x] = [self.grid[y][x]]

    def move_lose(self, loser):
        y, x, d, s = self.players[loser]

        for _ in range(4):
            dy, dx = self.directions[d]
            my, mx = y + dy, x + dx
            if not (1 <= my <= self.n and 1 <= mx <= self.n) or [i for i in self.players if i[:2] == [my, mx]]:
                d = (d + 1) % 4
            else:
                self.players[loser] = [my, mx, d, s]
                if len(self.gun_grid[my][mx]) > 0:
                    put_gun = self.gun[loser]
                    self.gun_grid[my][mx].sort()
                    self.gun[loser] = self.gun_grid[my][mx].pop()
                    if put_gun:
                        self.gun_grid[my][mx].append(put_gun)
                break

    def after_fight(self, winner, loser, y, x):
        if self.gun[loser] != 0:
            self.gun_grid[y][x].append(self.gun[loser])
            self.gun[loser] = 0  # 패자는 총 버림
            self.gun_grid[y][x].sort()
        self.move_lose(loser)  # 패자 이동

        if len(self.gun_grid[y][x]) > 0 and self.gun[winner] < self.gun_grid[y][x][-1]:  # 바닥에 총이 내거보다 좋으면 교환
            put_gun = self.gun[winner]
            self.gun[winner] = self.gun_grid[y][x].pop()
            self.gun_grid[y][x].append(put_gun)
            self.gun_grid[y][x].sort()
            # 소트

    def fight(self, player, enemy, y, x):
        player_atk = self.players[player][3] + self.gun[player]
        enemy_atk = self.players[enemy][3] + self.gun[enemy]
        if player_atk > enemy_atk:  # 주인공 승리
            self.points[player] += abs(player_atk - enemy_atk)
            self.after_fight(player, enemy, y, x)
        elif player_atk == enemy_atk:
            if self.players[player][3] > self.players[enemy][3]:  # 주인공 승리
                self.points[player] += abs(player_atk - enemy_atk)
                self.after_fight(player, enemy, y, x)
            elif self.players[player][3] < self.players[enemy][3]:  # 상대방 승리
                self.points[enemy] += abs(player_atk - enemy_atk)
                self.after_fight(enemy, player, y, x)
        else:  # 상대방 승리
            self.points[enemy] += abs(player_atk - enemy_atk)
            self.after_fight(enemy, player, y, x)

    def get_gun(self, player, my, mx):
        self.gun_grid[my][mx].sort()  # 가장 강한 총을 줍기 위해
        if len(self.gun_grid[my][mx]) > 0:  # 총이 바닥에 있고
            if self.gun[player] == 0:  # 내가 총이 없다면
                self.gun[player] = self.gun_grid[my][mx].pop()
            else:
                if self.gun[player] < self.gun_grid[my][mx][-1]:
                    put_gun = self.gun[player]
                    self.gun[player] = self.gun_grid[my][mx].pop()
                    self.gun_grid[my][mx].append(put_gun)

    def move(self):
        for player in range(1, self.m + 1):
            y, x, d, s = self.players[player]
            dy, dx = self.directions[d]
            my, mx = y + dy, x + dx
            if not (1 <= my <= self.n and 1 <= mx <= self.n):  # 벽에 부딪히면
                my, mx = y - dy, x - dx
                d = (d + 2) % 4
            self.players[player] = [my, mx, d, s]

            for enemy in range(1, self.m + 1):
                if player == enemy:  # 나는 범위에서 제외
                    continue
                if [my, mx] == self.players[enemy][:2]:  # 적이 있다면
                    self.fight(player, enemy, my, mx)
                    break
            else:  # 적이 없다면
                self.get_gun(player, my, mx)

    def solve(self):
        self.init_grid()
        for stage in range(1, self.k + 1):
            self.move()
            # print(self.gun[1:])
            # print(self.players)
            # print(*self.points[1:])

        print(*self.points[1:])


problem = Main()
problem.solve()

File: 241202/싸움땅/battle-ground.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int n, m, k;
	static int[][] map;
	static HashMap<String, PriorityQueue<Integer>> guns;
	static Player[] players;
	static int y, x, d, s;
	static int[][] directions = { { -1, 0 }, { 0, 1 }, { 1, 0 }, { 0, -1 } };

	static class Player {
		int y, x, d, s, gun, score;

		public Player(int y, int x, int d, int s, int gun, int score) {
			this.y = y;
			this.x = x;
			this.d = d;
			this.s = s;
			this.gun = gun;
			this.score = score;
		}

		@Override
		public String toString() {
			return "Player [y=" + y + ", x=" + x + ", d=" + d + ", s=" + s + ", gun=" + gun + ", score=" + score + "]";
		}
		
		
	}

	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		map = new int[n][n];
		guns = new HashMap<>();
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < n; j++) {
				String key = i + ", " + j;
				map[i][j] = Integer.parseInt(st.nextToken());
				guns.put(key, new PriorityQueue<>((o1, o2) -> o2 - o1));
				guns.get(key).add(map[i][j]);
			}
		}
		players = new Player[m];
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());
			y = Integer.parseInt(st.nextToken()) - 1;
			x = Integer.parseInt(st.nextToken()) - 1;
			d = Integer.parseInt(st.nextToken());
			s = Integer.parseInt(st.nextToken());
			players[i] = new Player(y, x, d, s, 0, 0);
		}
		
		solve();
	}
	
	static void move_loser(Player loser) {
		for(int i = 0; i < 4; i++) {
			int[] d = directions[loser.d];
			boolean flag = false;
			int my = loser.y + d[0];
			int mx = loser.x + d[1];
			boolean range = my < 0 || my >= n || mx < 0 || mx >= n;
			if(range) {
				loser.d = (loser.d + 1) % 4;
				continue;
			}
			for(Player player: players) {
				if(my == player.y && mx == player.x) {
					loser.d = (loser.d + 1) % 4;
					flag = true;
					break;
				}
			}
			if(flag) {
				continue;
			}
			loser.y = my;
			loser.x = mx;
			if(guns.get(loser.y + ", " + loser.x).peek() > 0) {
				int put_gun = loser.gun;
				loser.gun = guns.get(loser.y + ", " + loser.x).poll();
				if(put_gun > 0) {
					guns.get(loser.y + ", " + loser.x).add(put_gun);
				}
			}
			break;
		}
	}
	
	static void fight_player(Player player, Player enemy) {
		int player_atk = player.s + player.gun;
		int enemy_atk = enemy.s + enemy.gun;
		if(player_atk > enemy_atk) {
			player.score += Math.abs(player_atk - enemy_atk);
			if(enemy.gun > 0) {
				guns.get(enemy.y + ", " + enemy.x).add(enemy.gun);
				enemy.gun = 0;
			}
			get_gun(player);
			move_loser(enemy);
		} else if(player_atk == enemy_atk) {
			if(player.s > enemy.s) {
				player.score += Math.abs(player_atk - enemy_atk);
				if(enemy.gun > 0) {
					guns.get(enemy.y + ", " + enemy.x).add(enemy.gun);
					enemy.gun = 0;
				}
				get_gun(player);
				move_loser(enemy);
			} else if(player.s < enemy.s) {
				enemy.score += Math.abs(player_atk - enemy_atk);
				if(player.gun > 0) {
					guns.get(player.y + ", " + player.x).add(player.gun);
					player.gun = 0;
				}
				get_gun(enemy);
				move_loser(player);
			}
		} else {
			enemy.score += Math.abs(player_atk - enemy_atk);
			if(player.gun > 0) {
				guns.get(player.y + ", " + player.x).add(player.gun);
				player.gun = 0;
			}
			get_gun(enemy);
			move_loser(player);
		}
	}
	
	static void get_gun(Player player) {
		int put_gun = player.gun;
		String pos = player.y + ", " + player.x;
		if(!guns.get(pos).isEmpty() && guns.get(pos).peek() > put_gun) {
			player.gun = guns.get(pos).poll();
			guns.get(pos).add(put_gun);
		}
	}

	static void move_player() {
		for (int player = 0; player < m; player++) {
			boolean fight = false;
			int[] d = directions[players[player].d];
			int my = players[player].y + d[0];
			int mx = players[player].x + d[1];
			if (my < 0 || my >= n || mx < 0 || mx >= n) {
				my = players[player].y - d[0];
				mx = players[player].x - d[1];
				players[player].d = (players[player].d + 2) % 4;
			}
			players[player].y = my;
			players[player].x = mx;
			for(int enemy = 0; enemy < m; enemy++) {
				if(player == enemy) {
					continue;
				}
				if(players[player].y == players[enemy].y && players[player].x == players[enemy].x) {
					fight_player(players[player], players[enemy]);
					fight = true;
					break;
				}
			}
			if(!fight) {
				get_gun(players[player]);
			}
		}
	}

	static void solve() {
		for (int turn = 1; turn <= k; turn++) {
			move_player();
		}
		StringBuilder sb = new StringBuilder();
		for(Player player: players)
			sb.append(player.score).append(" ");
		System.out.println(sb);
	}
}
